File: docs/topic-memory-graph-main/src/inject.ts
/**
 * Replace or append a bounded section in markdown using HTML comment markers.
 * Safe for AGENTS.md-style files: markers are invisible in most Markdown renderers.
 */

export interface MemorySectionMarkers {
  readonly start: string;
  readonly end: string;
}

/** Neutral defaults for new integrations */
export const DEFAULT_MARKERS: MemorySectionMarkers = {
  start: "<!-- topic-memory-graph-start -->",
  end: "<!-- topic-memory-graph-end -->",
};

/** Markers used by DidClaw desktop (Tauri) — keep in sync when migrating hosts */
export const DIDCLAW_PHEROMONE_MARKERS: MemorySectionMarkers = {
  start: "<!-- didclaw-pheromone-start -->",
  end: "<!-- didclaw-pheromone-end -->",
};

/**
 * Insert `content` between markers in `existing`. If markers are absent, appends a new block.
 */
export function injectMemorySection(
  existing: string,
  content: string,
  markers: MemorySectionMarkers = DEFAULT_MARKERS,
): string {
  const { start, end } = markers;
  const section = `${start}\n${content}\n${end}`;

  const si = existing.indexOf(start);
  const ei = existing.indexOf(end);
  if (si !== -1 && ei !== -1 && ei > si) {
    const afterEnd = ei + end.length;
    return `${existing.slice(0, si)}${section}${existing.slice(afterEnd)}`;
  }

  const trimmed = existing.trimEnd();
  if (trimmed.length === 0) return section;
  return `${trimmed}\n\n${section}`;
}

File: crates/topic-memory/src/inject.rs
//! HTML-comment bounded injection markers.

#[derive(Debug, Clone, Copy)]
pub struct MemorySectionMarkers {
    pub start: &'static str,
    pub end: &'static str,
}

pub const DEFAULT_MARKERS: MemorySectionMarkers = MemorySectionMarkers {
    start: "<!-- topic-memory-graph-start -->",
    end: "<!-- topic-memory-graph-end -->",
};

/// Insert or replace a bounded memory section in markdown.
#[must_use]
pub fn inject_memory_section(
    existing: &str,
    content: &str,
    markers: MemorySectionMarkers,
) -> String {
    let section = format!("{}\n{}\n{}", markers.start, content, markers.end);
    let si = existing.find(markers.start);
    let ei = existing.find(markers.end);
    if let (Some(si), Some(ei)) = (si, ei)
        && ei > si
    {
        let after_end = ei + markers.end.len();
        return format!("{}{}{}", &existing[..si], section, &existing[after_end..]);
    }
    let trimmed = existing.trim_end();
    if trimmed.is_empty() {
        return section;
    }
    format!("{trimmed}\n\n{section}")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn replaces_existing_section() {
        let base = format!(
            "# Agents\n\n{}\nold\n{}\n",
            DEFAULT_MARKERS.start, DEFAULT_MARKERS.end
        );
        let out = inject_memory_section(&base, "new body", DEFAULT_MARKERS);
        assert!(out.contains("new body"));
        assert!(!out.contains("old"));
    }
}
